File: RL_DIDP_TSP/code/NQBH_TSP.cpp
#include <bits/stdc++.h>
/*
#include <iostream>
#include <vector>
#include <bitset>
#include <unordered_map>
#include <algorithm>
*/
using namespace std;
using ll = long long;

/* 1st naive solution
source: Geeks for Geeks/Traveling Salesman Problem (TSP) Implementation
url: https://www.geeksforgeeks.org/dsa/traveling-salesman-problem-tsp-implementation/
C++ program to find the shortest possible route that visits every city exactly once & returns to the starting point
time complexity: O(n!) since the algorithm uses the next_permutation function which generates all the possible permutations of the vertex set.
auxiliary space: O(n) since we use a vector to store all the vertices.
*/

int tsp(vector<vector<int>> &cost) {
	int num_nodes = cost.size(); // number of nodes
	vector<int> nodes;
	for (int i = 1; i < num_nodes; ++i) nodes.push_back(i); // initialize the nodes excluding the fixed starting point (node 0)
	int min_cost = INT_MAX;	
	do { // generate all permutations of the remaining nodes
		int curr_cost = 0, curr_node = 0; // start from node 0		
		for (int i = 0; i < (int)nodes.size(); ++i) { // calculate the cost of the current permutation
			curr_cost += cost[curr_node][nodes[i]];
			curr_node = nodes[i];
		}
		curr_cost += cost[curr_node][0]; // add the cost to return to the starting node
		min_cost = min(min_cost, curr_cost); // update the minimum cost if the current cost is lower
	} while (next_permutation(nodes.begin(), nodes.end()));
	return min_cost;
}

/* 2nd naive solution
source: Geeks for Geeks/Traveling Salesman Problem
url: https://www.geeksforgeeks.org/dsa/travelling-salesman-problem-using-dynamic-programming/
Naive approach: use DFS: O(n!) time & O(n) space
*/

int dfs(vector<vector<int>>& cost, vector<bool>& vis, int last, int cnt) {
	int n = cost.size();
	if (cnt == n) return cost[last][0]; // if all visited then return cost back to start 0,
	int min_cost = INT_MAX;
	for (int city = 1; city < n; ++city)
		if (!vis[city]) {
			vis[city] = true; // mark the city as visited & explore all possible paths from there
			min_cost = min(min_cost, cost[last][city] + dfs(cost, vis, city, cnt + 1));
			vis[city] = false; // backtrack
		}
	return min_cost; // return minimum cost among all possible paths
}

int tsp_dfs(vector<vector<int>>& cost) {
	int n = cost.size();
	vector<bool> vis(n, false); // guarantee that each city is visited exactly once
	vis[0] = true;
	return dfs(cost, vis, 0, 1);
}

/* 3rd solution: backtracking O(n!) time & O(n) space
source: Geeks for Geeks/Traveling Salesman Problem implementation using backtracking
url: www.geeksforgeeks.org/dsa/travelling-salesman-problem-implementation-using-backtracking/
C++ program to find the shortest possible route that visits every city exactly once & returns to the starting point using backtracking
*/

void total_cost_backtrack(vector<vector<int>> &cost, vector<bool> &visited, int curr_pos, int n, int cnt, int cost_so_far, int &ans) {
	if (cnt == n && cost[curr_pos][0]) { // if all nodes are visited & there is an edge to starting node
		ans = min(ans, cost_so_far + cost[curr_pos][0]); // update the minimum cost
		return;
	}
	for (int i = 0; i < n; ++i) // try visiting each node from current position
		if (!visited[i] && cost[curr_pos][i]) {
			visited[i] = true; // if node is not visited & has an edge then mark as visited
			total_cost_backtrack(cost, visited, i, n, cnt + 1, cost_so_far + cost[curr_pos][i], ans);
			visited[i] = false;
		}
}

int tsp_backtrack(vector<vector<int>> &cost) {
	int n = cost.size();
	vector<bool> visited(n, false);
	visited[0] = true;
	int ans = INT_MAX;
	total_cost_backtrack(cost, visited, 0, n, 1, 0, ans);
	return ans;
}

/* 4th solution: bitmask representation for tracking visited cities
Use recursion O(n!) time & O(n) space
*/

int total_cost(int mask, int pos, vector<vector<int>>& cost) {
	int n = cost.size();
	if (mask == (1 << n) - 1) return cost[pos][0]; // base case: if all cities are visited, return the cost to return to the starting city 0
	int ans = INT_MAX;
	for (int i = 0; i < n; ++i) // try visiting every city that has not been visited yet
		if (!(mask & (1 << i))) ans = min(ans, cost[pos][i] + total_cost(mask | (1 << i), i, cost)); // if city is not visited, visit it & update the mask
	return ans;
}

int tsp_bitmask(vector<vector<int>>& cost) {
	int mask = 1, pos = 0; // start from city 0, & only city 0 is visited initially (mask = 1)
	return total_cost(mask, pos, cost);
}

/* 5th solution: use top-down dynamic programming (memoization)
O(n^2 * 2^n) time & O(n * 2^n) space
*/

int total_cost_dp_topdown(int mask, int curr, vector<vector<int>>& cost, vector<vector<int>>& dp) {
	int n = cost.size();
	if (mask == (1 << n) - 1) return cost[curr][0]; // base case: if all cities are visited, return the cost to return to the starting city 0
	if (dp[curr][mask] != -1) return dp[curr][mask]; // if the value has already been computed, return it from the DP table
	int ans = INT_MAX;
	for (int i = 0; i < n; ++i) // try visiting every city that has not been visited yet
		if (!(mask & (1 << i))) ans = min(ans, cost[curr][i] + total_cost_dp_topdown(mask | (1 << i), i, cost, dp)); // if city is not visited, visity city i & update the mask
	return dp[curr][mask] = ans;
}

int tsp_dp_topdown(vector<vector<int>>& cost) {
	int n = cost.size();
	vector<vector<int>> dp(n, vector<int>(1 << n, -1));
	int mask = 1, curr = 0; // start from city 0, with only city 0 (visited initially, mask = 1)
	return total_cost_dp_topdown(mask, curr, cost, dp);
}

/* 6th solution: iterative dp, use bottom-up dynamic programming (tabulation)
O(n^2 * 2^n) time & O(n * 2^n) space
*/
int tsp_bottomup(vector<vector<int>>& cost) {
	int n = cost.size();
	if (n <= 1) return n == 1 ? cost[0][0] : 0;
	const int INF = INT_MAX; // maximum cost to visit all cities
	int FULL = 1 << n, full_mask = FULL - 1;
	vector<vector<int>> dp(FULL, vector<int>(n, INF)); // dp[mask][i] represents the minimum cost to visit all cities corresponding to the set bits in mask, ending at city i
	dp[1][0] = 0;
	for (int mask = 1; mask < FULL; ++mask) // iterate over all subsets of cities
		for (int i = 0; i < n; ++i) {
			// skip if city i is not included in mask
			if (!(mask & (1 << i))) continue;
			if (dp[mask][i] == INF) continue;
			for (int j = 0; j < n; ++j) { // try to go to every unvisited city j
				if (mask & ( 1 << j)) continue; // skip if city j is already visited
				// cost to visit new city j from city i s.t. previously visited cities remain visited
				int nxt = mask | (1 << j);
				dp[nxt][j] = min(dp[nxt][j], dp[mask][i] + cost[i][j]);
			}
		}
	int ans = INF;
	for (int i = 0; i < n; ++i)
		if (dp[full_mask][i] != INF) // if last city on path is i & cost of path is not infinity
			ans = min(ans, dp[full_mask][i] + cost[i][0]); // update net cost s.t. city 0 is visited in last
	return ans;
}

/* implementation a Reinforcement-Guided Domain-Independent Dynamic Programming (RG-DIDP) solver in C++, we focus on a structure that allows an external model (the RL agent) to inject heuristic values into a high-performance search engine.
The following architecture uses a Beam Search approach, as it is the standard for memory-efficient DIDP.
Core State Representation: In C++, efficiency is gained by using bitmasks for the set of visited cities and custom hash functions for the DP table.
*/

struct state {
	int curr_city;
	std::bitset<32> visited; // assuming TSP size <= 32 for bitmask
	double g_cost; // path cost from start to here
	bool operator==(const state& other) const {
		return curr_city == other.curr_city && visited == other.visited;
	}
};

// custom hash for DP table
struct state_hash {
	std::size_t operator()(const state& s) const {
        return std::hash<int>()(s.curr_city) ^ (std::hash<std::bitset<32>>()(s.visited) << 1);
    }
};

/*
// Integration with the RL Model: show how the C++ engine communicates with the RL model (typically trained in Python/PyTorch). We use LibTorch (the PyTorch C++ API) for low-latency inference during the search.
#include <torch/script.h> // LibTorch

class RLheuristic {
private:
	torch::jit::script::Module model;
public:
	RLHeuristic(std::string model_path) {
        model = torch::jit::load(model_path);
    }

    double get_h_value(const state& s, const vector<vector<double>>& adj_matrix) {
    	auto tensor_state = state_to_tensor(s); // convert state & adjacency matrix to Torch Tensors
    	auto output = model.forward({tensor_state}).toTensor();
    	return output.item<double>();
    }
};

// RG-DIDP Search Loop: This is where the "Reinforcement-Guided" logic prunes the state space. We use a priority_queue to manage the "Beam" of states.
void solve_RL_DIDP(int n_cities, const vector<vector<double>>& dist) {
	unordered_map<state, double, state_hash> dp_table;
	vector<state> beam;

	// initial state
	state start = {0, bitset<32>(1), 0.0};
	beam.push_back(start);

	for (int step = 1; step < n_cities; ++step) {
		vector<state> next_beam;
		for (const auto& curr : beam)
			for (int next_city = 0; next_city < n_cities; ++next_city)
				if (!curr.visited.test(next_city)) {
					state next_s = {next_city, current.visited, current.g_cost + dist[curr.curr_city][next_city]};
					next_s.visited.set(next_city);
					next_beam.push_back(next_s);
				}
		// Reinforcement Learning guidance step
		// sort by f = g + h_RL
		sort(next_beam.begin(), next_beam.end(), [&](const state& a, const state& b) {
			double f_a = a.g_cost + rl_agent.get_h_value(a, dist);
            double f_b = b.g_cost + rl_agent.get_h_value(b, dist);
            return f_a < f_b;
		});
		// prune: keep only the top 'W' (Beam Width) states
		if (next_beam.size() > BEAM_WIDTH) next_beam.resize(BEAM_WIDTH);
		beam = next_beam;
	}
}
*/

//-----------------------------------------------------------------------------//
// main
//-----------------------------------------------------------------------------//

int main() {
	// small fixed testcase
	vector<vector<int>> cost = {{0, 10, 15, 20, 12}, 
                                {10, 0, 35, 25, 21}, 
                                {15, 35, 0, 30, 7}, 
                                {20, 25, 30, 0, 9},
                            	{23, 37, 57, 19, 0}};
    // input testcase

    // output result
    cout << tsp(cost) << '\n';
    cout << tsp_dfs(cost) << '\n';
    cout << tsp_backtrack(cost) << '\n';
    cout << tsp_bitmask(cost) << '\n';
    cout << tsp_dp_topdown(cost) << '\n';
    cout << tsp_bottomup(cost) << '\n';
	/*
    torch::Tensor tensor = torch::rand({2, 3});
    std::cout << "Random Tensor: " << tensor << std::endl;
    */
}

File: RL_DIDP_TSP/code/NQBH_TSP.py
import time
import torch
import torch.nn as nn
import numpy as np
from dataclasses import dataclass
from typing import List
import torch.nn.functional as F

"""
Method 1: Dynamic Programming (Held-Karp Algorithm)
This is the standard algorithmic solution. It uses bitmasking to keep track of visited cities and recursion with memoization to avoid re-calculating the same sub-paths.
Time Complexity: O(n^2 * 2^n) Space Complexity: O(n * 2^n)
"""
import sys

def solve_tsp_dynamic(dist_matrix):
    """
    Solves TSP using Held-Karp algorithm (Dynamic Programming).
    Returns (min_cost, path).
    """
    n = len(dist_matrix)
    
    # memoization table: key=(mask, current_city), value=min_cost
    memo = {}
    # parent table to reconstruct the path: key=(mask, current_city), value=prev_city
    parent = {}

    # State: mask (bitmask of visited cities), pos (current city index)
    def visit(mask, pos):
        # Base case: All cities visited? Return distance to return to start (city 0)
        if mask == (1 << n) - 1:
            return dist_matrix[pos][0] or float('inf') # Return inf if no path back

        state = (mask, pos)
        if state in memo:
            return memo[state]

        ans = float('inf')
        best_next_city = -1

        # Try visiting every unvisited city
        for city in range(n):
            if (mask >> city) & 1 == 0:  # If city not in mask
                new_ans = dist_matrix[pos][city] + visit(mask | (1 << city), city)
                if new_ans < ans:
                    ans = new_ans
                    best_next_city = city
        
        memo[state] = ans
        parent[state] = best_next_city
        return ans

    # Start from city 0, with only city 0 in the mask (1 << 0)
    min_cost = visit(1, 0)
    
    # --- Path Reconstruction ---
    path = [0]
    curr_mask = 1
    curr_city = 0
    
    # We trace the path until we have visited all cities
    for _ in range(n - 1):
        next_city = parent.get((curr_mask, curr_city))
        path.append(next_city)
        curr_mask |= (1 << next_city)
        curr_city = next_city
        
    path.append(0) # Return to start
    
    return min_cost, path

"""
Method 2: Brute Force (Permutations)
If your dataset is very small (n <= 10), this code is much easier to read and debug. It simply generates every possible order of cities and calculates the total distance.
"""

from itertools import permutations

def solve_tsp_brute_force(dist_matrix):
    n = len(dist_matrix)
    cities = list(range(n))
    min_path = None
    min_dist = float('inf')

    # Try every permutation of cities (starting at 0)
    # We fix the start node (0) to reduce permutations by factor of n
    for perm in permutations(cities[1:]):
        current_path = [0] + list(perm) + [0]
        current_dist = 0
        
        # Calculate distance for this path
        for i in range(len(current_path) - 1):
            u, v = current_path[i], current_path[i+1]
            current_dist += dist_matrix[u][v]
        
        if current_dist < min_dist:
            min_dist = current_dist
            min_path = current_path

    return min_dist, min_path

"""
Method 3: RL-guided DIDP
"""

# 1. State and Model Definitions
@dataclass(frozen=True)
class State:
    current_city: int
    visited_mask: int
    g_cost: float

class TSPGuidanceGNN(nn.Module):
    def __init__(self, node_dim=3, hidden_dim=64):
        super().__init__()
        self.fc = nn.Sequential(
            nn.Linear(node_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )

    def forward(self, x):
        # Global mean pooling for simplicity in this example
        graph_latent = torch.mean(x, dim=1) 
        return self.fc(graph_latent)

# 2. Solver Class
class RGDIDPSolver:
    def __init__(self, model: nn.Module, beam_width: int):
        self.model = model
        self.beam_width = beam_width

    def get_h_batch(self, states: List[State], coords: np.ndarray):
        num_cities = len(coords)
        batch_features = []
        for s in states:
            # Build feature: [x, y, visited_status]
            visited = np.array([(s.visited_mask >> i) & 1 for i in range(num_cities)])
            node_feat = np.column_stack((coords, visited))
            batch_features.append(node_feat)
        
        inputs = torch.tensor(np.array(batch_features), dtype=torch.float32)
        with torch.no_grad():
            return self.model(inputs).squeeze().numpy()

    def solve(self, dist_matrix: np.ndarray, coords: np.ndarray):
        num_cities = len(dist_matrix)
        # Initial State: Start at City 0
        beam = [State(0, 1 << 0, 0.0)]

        for _ in range(1, num_cities):
            next_candidates = []
            for state in beam:
                for next_city in range(num_cities):
                    if not (state.visited_mask & (1 << next_city)):
                        new_g = state.g_cost + dist_matrix[state.current_city][next_city]
                        new_mask = state.visited_mask | (1 << next_city)
                        next_candidates.append(State(next_city, new_mask, new_g))

            if not next_candidates: break
            
            # Batch Heuristic Calculation
            h_scores = self.get_h_batch(next_candidates, coords)
            
            # Sorting & Pruning
            scored = sorted(zip(next_candidates, h_scores), 
                            key=lambda x: x[0].g_cost + x[1])
            beam = [item[0] for item in scored[:self.beam_width]]

        # Final return to origin
        return min(s.g_cost + dist_matrix[s.current_city][0] for s in beam)

# Training Algorithm: Fitted Value Iteration
# This script demonstrates how to generate trajectories and update the GNN.

import torch.optim as optim



# MAIN FUNCTION: Running an Instance
if __name__ == "__main__":
    # Setup dummy data for 5 cities
    num_cities = 12 # >= 13: long runtime
    coords = np.random.rand(num_cities, 2)
    dist_matrix = np.sqrt(((coords[:, None, :] - coords[None, :, :])**2).sum(-1))
    # dist_matrix = [[0, 10, 15, 20], [10, 0, 35, 25], [15, 35, 0, 30], [20, 25, 30, 0]];

    # Initialize Model and Solver
    start_time = time.perf_counter()
    
    model = TSPGuidanceGNN()
    solver = RGDIDPSolver(model, beam_width=5)    
    final_cost = solver.solve(dist_matrix, coords)
    print(f"Final Optimized Tour Cost: {final_cost}")

    end_time = time.perf_counter()
    elapsed_time = end_time - start_time
    print(f"Time elapsed: {elapsed_time} seconds")

    # --- Example Usage --- Comparision
    # Distance matrix (Adjacency Matrix)
    # Example: 5 cities
    start_time = time.perf_counter()
    cost, path = solve_tsp_dynamic(dist_matrix)
    print(f"Minimum Cost: {cost}")
    print(f"Optimal Path: {path}")

    end_time = time.perf_counter()
    elapsed_time = end_time - start_time
    print(f"Time elapsed: {elapsed_time} seconds")

    start_time = time.perf_counter()
    cost, path = solve_tsp_brute_force(dist_matrix)
    print(f"Brute Force Cost: {cost}")
    print(f"Brute Force Path: {path}")

    end_time = time.perf_counter()
    elapsed_time = end_time - start_time
    print(f"Time elapsed: {elapsed_time} seconds")